cirq/google/serializable_device.py

```diff
@@ -0,0 +1,136 @@
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class SerializableDevice(devices.Device):
+    """Device object generated from a device specification proto.
+
+    Given a device specification proto and a gate_set to translate the
+    serialized gate_ids to cirq Gates, this will generate a Device that can
+    verify operations and circuits for the hardware specified by the device.
+    """
+
+    def __init__(self, proto: device_pb2.DeviceSpecification,
+                 gate_set: serializable_gate_set.SerializableGateSet):
+        """
+
+        Args:
+            proto: A proto describing the qubits on the device, as well as the
+                supported gates and timing information.
+            gate_set: A SerializableGateSet that can translate the gate_ids
+                into cirq Gates.
+        """
+
+        self.qubits = self._qubits_from_ids(proto.valid_qubits)
+        self.allowed_targets: Dict[str, Set[Tuple]] = dict()
+        for ts in proto.valid_targets:
+            self.allowed_targets[ts.name] = self._create_target_set(ts)
+
+        gate_defs: Dict[str, device_pb2.GateDefinition] = dict()
+        for gs in proto.valid_gate_sets:
+            for gate_def in gs.valid_gates:
+                gate_defs[gate_def.id] = gate_def
+
+        self.durations: Dict[Any, Duration] = dict()
+        self.target_sets: Dict[Any, List[str]] = dict()
+        for gate_type in gate_set.supported_gate_types():
+            for serializer in gate_set.serializers[gate_type]:
+                gate_id = serializer.serialized_gate_id
+                if gate_id not in gate_defs:
+                    raise ValueError(f'Serializer has {gate_id} which is not ' +
+                                     'supported by the device specification')
+                gate_picos = gate_defs[gate_id].gate_duration_picos
+                self.durations[gate_type] = Duration(picos=gate_picos)
+                self.target_sets[gate_type] = gate_defs[gate_id].valid_targets
+        print(self.durations)
+
+    def _qubits_from_ids(self, id_list):
+        """Translates a list of ids in proto format e.g. '4_3'
+        into cirq.GridQubit objects"""
+        # TODO(dstrain): Add support for non-grid qubits
+        return [devices.GridQubit.from_proto_id(id) for id in id_list]
+
+    def _create_target_set(self, ts: device_pb2.TargetSet) -> Set[Tuple]:
+        """Transform a TargetSet proto into a set of qubit tuples"""
+        # TODO(dstrain): add support for measurement qubits
+        target_set = set()
+        for target in ts.targets:
+            qid_list = self._qubits_from_ids(target.ids)
+            target_set.add(tuple(qid_list))
+            if ts.target_ordering == device_pb2.TargetSet.SYMMETRIC:
+                qid_list.reverse()
+                target_set.add(tuple(qid_list))
+        return target_set
+
+    def _find_operation_type(self, op_key: 'cirq.Operation',
+                             type_dict: Dict[Any, Any]):
```

Comment:
can this type annotation be made more specific? the docstring says the keys have `Gate` type

Reply:
Done.